### PR TITLE
Better handling of selected contours in cvstroke.c

### PR DESCRIPTION
**Pull request by skef**
@jtanx: With respect to https://github.com/fontforge/fontforge/pull/3991#issuecomment-553302278, wouldn't the enclosed changes account for the problem and make more sense?

### CODE REVIEW

**Comment by skef** [comment]
@frank-trampe You may remember a while back I said I was seeing crashes in RemoveOverlap and then a bit later (by email) I said:

> 1. The problems don't arise now when using the python interface.
> 2. Thinking about it, I can't remember encountering the problems when using the python interface.
>
> Thinking about it still more, this isn't necessarily all that mysterious to me. I was already aware of a relation in the GUI (which is, perhaps strangely, where I've been doing most of my testing) between patterns of point selection and crashes. That stems from CVStrokeIt() attending to which contours are "selected", and my trying to accommodate that with some code that ultimately calls SplineSetRemoveOverlap() with over_rmselected.

In retrospect it seems as though the code changed in this PR was the source of that problem.

**Comment by frank-trampe** [comment]
So the problem was that `prev` was getting set to the last element rather than the first?

**Comment by skef** [comment]
The loop is supposed to go through each contour in the layer and:

1. If it has selected points, stroke it and add the produced contours back into the layer
2. If it doesn't have selected points, just add it back into the layer. 

The contour in question at a given loop iteration is `spl`. `snext` is a means of keeping track of the next element in the list for the next iteration. 

The existing code also adds `snext` back into the list. This is just a strange mistake, as far as I can tell.

**Comment by skef** [comment]
(I assume that I suspected Remove Overlap because of the combination of deleted contours being added to the layer and that facility's energetic allocation of memory. Most of the rest of the code I was testing with doesn't allocate much and probably left the freed structures intact.)

**Comment by jtanx** [comment]
This still causes issues with asan

```
==1110==ERROR: AddressSanitizer: heap-use-after-free on address 0x60800000f928 at pc 0x0000005ce4e9 bp 0x7ffd67b25050 sp 0x7ffd67b25048
READ of size 8 at 0x60800000f928 thread T0
    #0 0x5ce4e8 in CVInfoDrawText /home/jeremy/ffstaging/build/../fontforgeexe/charview.c
    #1 0x61614c in CVInfoDraw /home/jeremy/ffstaging/build/../fontforgeexe/charview.c:4171:5
    #2 0x61614c in CVCrossing /home/jeremy/ffstaging/build/../fontforgeexe/charview.c:4182
    #3 0x61614c in v_e_h /home/jeremy/ffstaging/build/../fontforgeexe/charview.c:5719
    #4 0xa22872 in GiveToAll /home/jeremy/ffstaging/build/../gdraw/gcontainer.c:588:6
    #5 0xa2025a in _GWidget_TopLevel_eh /home/jeremy/ffstaging/build/../gdraw/gcontainer.c:679:2
    #6 0xa47a1d in _GGDKDraw_CallEHChecked /home/jeremy/ffstaging/build/../gdraw/ggdkdraw.c:334:9
    #7 0xa47a1d in _GGDKDraw_DispatchEvent /home/jeremy/ffstaging/build/../gdraw/ggdkdraw.c:1237
    #8 0x7fcf13c5b764  (/usr/lib/x86_64-linux-gnu/libgdk-3.so.0+0x37764)
    #9 0x7fcf13c8bf91  (/usr/lib/x86_64-linux-gnu/libgdk-3.so.0+0x67f91)
    #10 0x7fcf12b61416 in g_main_context_dispatch (/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x4c416)
    #11 0x7fcf12b6164f  (/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x4c64f)
    #12 0x7fcf12b616db in g_main_context_iteration (/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x4c6db)
    #13 0xa518a9 in GGDKDrawProcessPendingEvents /home/jeremy/ffstaging/build/../gdraw/ggdkdraw.c:2176:16
    #14 0x6f933d in MakeStrokeDlg /home/jeremy/ffstaging/build/../fontforgeexe/cvstroke.c:1212:9
    #15 0xa8906e in GMenuBarCheckHotkey /home/jeremy/ffstaging/build/../gdraw/gmenu.c:1966:7
    #16 0xa8906e in GMenuBarCheckKey /home/jeremy/ffstaging/build/../gdraw/gmenu.c:2015
    #17 0xa21d5f in _GWidget_TopLevel_Key /home/jeremy/ffstaging/build/../gdraw/gcontainer.c:507:23
    #18 0xa20277 in _GWidget_TopLevel_eh /home/jeremy/ffstaging/build/../gdraw/gcontainer.c:682:9
    #19 0xa47a1d in _GGDKDraw_CallEHChecked /home/jeremy/ffstaging/build/../gdraw/ggdkdraw.c:334:9
    #20 0xa47a1d in _GGDKDraw_DispatchEvent /home/jeremy/ffstaging/build/../gdraw/ggdkdraw.c:1237
    #21 0x7fcf13c5b764  (/usr/lib/x86_64-linux-gnu/libgdk-3.so.0+0x37764)
    #22 0x7fcf13c8bf91  (/usr/lib/x86_64-linux-gnu/libgdk-3.so.0+0x67f91)
    #23 0x7fcf12b61416 in g_main_context_dispatch (/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x4c416)
    #24 0x7fcf12b6164f  (/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x4c64f)
    #25 0x7fcf12b616db in g_main_context_iteration (/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x4c6db)
    #26 0xa51aa8 in GGDKDrawEventLoop /home/jeremy/ffstaging/build/../gdraw/ggdkdraw.c:2204:17
    #27 0x9b5d03 in fontforge_main /home/jeremy/ffstaging/build/../fontforgeexe/startui.c:1415:5
    #28 0x7fcf12745b96 in __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:310
    #29 0x433ce9 in _start (/home/jeremy/ffstaging/build/bin/fontforge+0x433ce9)

0x60800000f928 is located 8 bytes inside of 96-byte region [0x60800000f920,0x60800000f980)
freed by thread T0 here:
    #0 0x4fdc38 in __interceptor_free (/home/jeremy/ffstaging/build/bin/fontforge+0x4fdc38)
    #1 0x7fcf1632b74c in SplinePointMDFree /home/jeremy/ffstaging/build/../fontforge/splineutil.c:212:5
    #2 0x7fcf1632b74c in SplinePointListMDFree /home/jeremy/ffstaging/build/../fontforge/splineutil.c:265
    #3 0x6fb7e5 in CVStrokeIt /home/jeremy/ffstaging/build/../fontforgeexe/cvstroke.c:125:3
    #4 0x707aa0 in _Stroke_OK /home/jeremy/ffstaging/build/../fontforgeexe/cvstroke.c:263:2
    #5 0x708e5b in Stroke_OK /home/jeremy/ffstaging/build/../fontforgeexe/cvstroke.c:273:9
    #6 0xa0d553 in GButtonInvoked /home/jeremy/ffstaging/build/../gdraw/gbuttons.c:257:2
    #7 0xa08305 in gbutton_mouse /home/jeremy/ffstaging/build/../gdraw/gbuttons.c:485:3
    #8 0xa1ef13 in _GWidget_Container_eh /home/jeremy/ffstaging/build/../gdraw/gcontainer.c:304:16
    #9 0xa21015 in _GWidget_TopLevel_eh /home/jeremy/ffstaging/build/../gdraw/gcontainer.c:747:11
    #10 0xa47a1d in _GGDKDraw_CallEHChecked /home/jeremy/ffstaging/build/../gdraw/ggdkdraw.c:334:9
    #11 0xa47a1d in _GGDKDraw_DispatchEvent /home/jeremy/ffstaging/build/../gdraw/ggdkdraw.c:1237
    #12 0x7fcf13c5b764  (/usr/lib/x86_64-linux-gnu/libgdk-3.so.0+0x37764)

previously allocated by thread T0 here:
    #0 0x4fe24f in calloc (/home/jeremy/ffstaging/build/bin/fontforge+0x4fe24f)
    #1 0x7fcf161ff902 in SFDGetSplineSet /home/jeremy/ffstaging/build/../fontforge/sfd.c:4126:8
    #2 0x7fcf1621d5bf in SFDGetChar /home/jeremy/ffstaging/build/../fontforge/sfd.c:5698:42
    #3 0x7fcf16222d54 in ModSF /home/jeremy/ffstaging/build/../fontforge/sfd.c:9301:19
    #4 0x7fcf16222d54 in SlurpRecovery /home/jeremy/ffstaging/build/../fontforge/sfd.c:9348
    #5 0x7fcf16222d54 in SFRecoverFile /home/jeremy/ffstaging/build/../fontforge/sfd.c:9434
    #6 0x7fcf15d8c04f in DoAutoRecoveryExtended /home/jeremy/ffstaging/build/../fontforge/autosave.c:111:13
    #7 0x9b559d in fontforge_main /home/jeremy/ffstaging/build/../fontforgeexe/startui.c:1314:8
    #8 0x7fcf12745b96 in __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:310

SUMMARY: AddressSanitizer: heap-use-after-free /home/jeremy/ffstaging/build/../fontforgeexe/charview.c in CVInfoDrawText
Shadow bytes around the buggy address:
  0x0c107fff9ed0: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c107fff9ee0: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c107fff9ef0: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c107fff9f00: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c107fff9f10: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c107fff9f20: fa fa fa fa fd[fd]fd fd fd fd fd fd fd fd fd fd
  0x0c107fff9f30: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c107fff9f40: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c107fff9f50: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c107fff9f60: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c107fff9f70: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==1110==ABORTING

```

File: fontforgeexe/cvstroke.c [hunk 1]
```diff
@@ -114,23 +114,23 @@ static void CVStrokeIt(void *_cv, StrokeInfo *si, int justapply) {
 		cur = SplineSetStroke(spl,si,cv->b.layerheads[cv->b.drawmode]->order2);
```
Comment (jtanx) [inline fontforgeexe/cvstroke.c]: Not quite sure if this has the right effect, but maybe we could insert a 

```c
SCClearSelPt(cv->b.sc);
```
somewhere around here.

Comment (skef) [reply fontforgeexe/cvstroke.c]: I don't think the stroke code itself cares at all about selection.

**Comment by skef** [comment]
@jtanx with that test are you using Apply at any point or just OK?

File: fontforgeexe/cvstroke.c (hunk 1 above)
Comment (jtanx) [reply fontforgeexe/cvstroke.c]: Yeah, but that's what's causing the valgrind/memory issues.

In this function you've just freed the SPL but there's still references to a selected point in that SPL, which is then being dereferenced in the charview (on cv->p)

**Comment by jtanx** [comment]
I just hit enter, so OK

**Comment by skef** [comment]
It's probably the lack of `spl->next = NULL` in the `else` section.

**Comment by skef** [comment]
Oh, we track selected points? I'm not sure I knew that.

**Comment by skef** [comment]
Can you just check moving the `spl->next` up to right after `snext` is set and confirm whether that does or doesn't work?

**Comment by jtanx** [comment]
That still has issues, like I said, there's references held into that freed SPL at `cv->p.sp`

**Comment by skef** [comment]
`CVCharChangedUpdate()` is supposed to take care of this.

**Comment by skef** [comment]
I guess it's unlikely the commit I just pushed will fix the problem, but at least the code makes a bit more sense now.

**Comment by skef** [comment]
Incidentally I spent some time trying to get address sanitizing working on my system but it became clear I didn't have enough CMake knowledge. I *should* make that intellectual investment, but for the time being do you have an easy means for this? Would it be hard to add an ASAN build_type (which is mentioned in some CMake docs but doesn't seem to work out of the box for FontForge)?

**Comment by jtanx** [comment]
Yeah I'll probably add something to make this easier to do, but fwiw this is what I do:

```
 CC=clang-7 CXX=clang++-7 CFLAGS="-g -fsanitize=address -fno-omit-frame-pointer" cmake .. -GNinja -DCMAKE_BUILD_TYPE=Debug -DENABLE_PYTHON_SCRIPTING=no -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DCMAKE_C_COMPILER_LAUNCHER=ccache
```

You probably don't have to use clang, but that's just what I had around at the time.

You could also forego all that and just use valgrind, although it's slower.

**Comment by skef** [comment]
Well, the last commit fixes the problem -- it wasn't `p.sp` (which was already cleared by `_CV_CharChangedUpdate()`) but `lastselpt`. Can you think of any reason not to clear the latter in this context?

**Comment by jtanx** [comment]
I'm fairly certain that isn't the right place to null it out, as _CV_CharChangedUpdate is called fairly liberally. Like I'm pretty sure it messes up trying to manipulate/move the control points as-is.

**Comment by skef** [comment]
`p.sp` and `lastselpt` are set together and checked together in a number of places, so the fact that the former is already nulled out in the function provides at least some evidence that it's appropriate to null out the latter. 

Weirdly, a lot of these lines with either (or both) of these variables are unchanged since 2001. Why would you *start out* with both of these structures rather than just using `pressedOn`? I guess the idea is to provide a record of the last selected point when something that isn't a point was pressed on more recently? 

In any case, it seems like any operation that deletes points should clear it.

**Comment by skef** [comment]
I guess `charview.c:_CVMerge()` calls both `SCClearSelPt()` and `CVCharChangedUpdate()`, so if you want to use that as the model for `CVStrokeIt()` I'm fine with that.

**Comment by skef** [comment]
I changed the code to do that. I think the fully virtuous investigation on this one can be left to a later time and developer.

**Comment by skef** [comment]
(Which was, reminding myself by looking back, your suggestion anyway.)

**Pull request merged**